README.md
# battery-monitoring-v1_2
Program Battery Monitoring (Node.js - Express - React). Kasus aplikasi versi ini: hardware level 0 AC Monitoring<br />
This project was bootstrapped with [Create React App](https://github.com/facebook/create-react-app).

## Setup Software
1. Clone (atau download zip)
2. Buka folder bernama depan arduino-, lalu upload kode tersebut ke Arduino. Cek COM port\* nya, dan sesuaikan pada langkah Setup untuk program sisi node.
3. Dengan terminal Powershell #1 yang terdapat di dalam VSCode, pada direktori root (yang ada README.md nya - seharusnya tidak perlu diubah bila folder ini yang dibuka di VSCode), ketik ```npm install```
4. Dengan terminal Powershell #2 (split atau new terminal), pergi ke direktori backend dengan ```cd backend```, lalu ```node node_server.js```
5. Kembali ke root pada terminal #1 dengan ```cd ../```, lalu ```npm start```
6. Cek di browser, http://localhost:3000 untuk UI React, http://localhost:5000/buffer untuk buffer data nya.<br />
    Sementara ini, program dibuat untuk single user. Login dengan username "root" dan password "password".

## Setup Hardware (Battery Monitoring)
1. Pastikan Arduino dengan Battery Monitoring pada kondisi yang sama dengan (\*), dengan baudRate 9600. Jika bukan di COM4, ganti bagian ini di file node_server.js:
    ```
    //deklarasi objek serial
    var portName = "COM3";
    var myPort = new serialport(portName,{
        baudRate:9600
    });
    ```
2. Sejauh ini cuma itu sih

## Setup Hardware (AC Monitoring)
1. Pastikan Arduino dengan AC Monitoring pada kondisi default ada di COM3, dengan baudRate 115200. 

## Menu yang Tersedia
1. Home <br /> Sebagai halaman yang memuat tampilan grafik keempat device, yang masing-masing memuat 8 group tag.
2. Konfigurasi Admin <br /> Sebagai halaman yang digunakan untuk mengubah konfigurasi tag-tag. Perlu login untuk mengakses halaman ini
3. Informasi Proyek <br /> Calon halaman untuk menempatkan detil seputar proyek
4. About Us <br /> Calon halaman "Tentang Kami"

## Link Router yang Digunakan
1. [Battery Monitoring](http://localhost:3000), sebagai aplikasi UI utama.
2. [Server Dummy](http://localhost:5000/api/buffer), sebagai tempat sementara data yang dikirim dari level 0
3. [Request Buffer](http://localhost:5000/api/request), sebagai tempat sementara untuk menampung permintaan perubahan konfigurasi dari Webserver.
4. [Request IO](http://localhost:5000/api/request_io), sebagai tempat sementara untuk menampung permintaan perubahan I/O dari Webserver.

## Changelog
Lihat CHANGELOG.md.

## Useful Sources: 
Silakan cek file DAFPUS.md untuk mengetahui sumber di balik trik-trik pemrograman yang telah dipelajari.

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.<br/>
**FORMAT TANGGAL : YYYY-MM-DD <br/>**

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

## [1.2] - 2020-11-20

### Added
- Panel I/O untuk 8 buah DI dan 8 buah AI yang diletakkan di Group Tag 1A milik Device 1. Diatur untuk menuliskan data ke Hardware AC monitoring. Penulisan data ke level 0 dilakukan jika data diubah saja, tidak setiap saat.
- Fasilitas untuk melakukan konfigurasi label, scaling, dan aksesibilitas dari tombol DI serta slider AI.
- Variabel konfigurasi dan data buffer milik DI dan AI generik dengan kapasitas 256 buah data. Pada versi ini, digunakan 8 item pertama untuk masing-masing input.
- Link Router baru, yaitu http://localhost:5000/api/request_io yang berfungsi untuk tempat buffer permintaan perubahan I/O dari webserver

### Changed
- Konfigurasi default sekarang disediakan dalam lebih dari 1 file. Harapannya, dengan copy-paste konfigurasi, use case dapat diganti dengan cepat.
- Tombol-tombol panel DI difungsikan 7 buah, slider panel AI belum difungsikan. Kedua jenis tersebut sudah di-generate untuk sisi Javascript.
- Bentuk variabel konfigurasi yang dimuat pada halaman Konfigurasi diubah dengan ditambahkannya konfigurasi I/O untuk setiap grouptag. Variabel ini dimuat sekaligus ketika halaman Home dimuat.
- Server Node (node_server.js) kini hanya memuat data kiriman dari arduino ke buffer jika ditandai dengan kode fungsi ```0x11```. Hal ini diterapkan untuk  mencegah terjadinya pembacaan yang tidak konsisten.

## [1.1] - 2020-11-02

### Added
- Generator buffer Dummy berupa integer random antara 800 - 1200, dan sebuah timestamp dengan format ```hh:mm:ss```. Mewakili struktur buffer baru yang dikirim dari node server dan diterima React Webserver.
- Halaman http://localhost:5000/api/request sebagai buffer komunikasi arah bawah (Webserver ke node server). Diisi bila terdapat perubahan yang dilakukan dari halaman Konfigurasi Admin.
- Fungsi untuk mengirim (HTTP POST) konfigurasi baru ke node server, dan menerima (HTTP GET) data buffer saat ini serta konfigurasi aktif saat ini.
- Tampilan pada halaman Home sekarang sudah memuat dua komponen : tabel berisi nama dan nilai saat ini, dan sebuah chart yang merekam perubahan nilai tersebut.
- Konektivitas dengan Arduino yang melakukan generate data dummy yang bentuknya sama dengan data Dummy.
- Fungsi scaling untuk data-data yang dimuat di chart, dan mapping berdasarkan indeks (0-255) dan device nya. Dari 5 device yang direncanakan, baru 2 yang dapat dijalankan di versi ini: Dummy dan Arduino.

### Changed
- Komponen-komponen HTML yang ada dalam halaman Konfigurasi Admin sekarang berubah menjadi Controlled Element (perubahan nilainya dipantau framework React), dan masing-masing komponen tadi adalah unik.
- Tulisan-tulisan nama Device dan Group Tag pada tampilan pada menu Home sekarang mengikuti konfigurasi apa yang aktif saat ini.
- Struktur file dan folder sudah dirapikan sedemikian rupa sehingga mengikuti kaidah (sementara) berikut:
    - folder ./src/components memuat komponen terkecil yang dapat dimuat dalam suatu page, atau elemen UI di dalam page.
    - folder ./src/config memuat konfigurasi default yang dimuat saat aplikasi pertama kali dijalankan
    - folder ./src/layouts memuat Primary Layout, yaitu layout utama aplikasi. Berisi 4 buah router (menu yang dapat diakses pengguna), dan wilayah kosong yang memuat konten menu terpilih.
    - folder ./src/pages  memuat halaman-halaman yang dimuat di dalam PrimaryLayout.
    - folder ./src/ui memuat elemen-elemen UI yang mungkin dimuat di dalam halaman tadi (bisa saja dalam satu halaman ada dua elemen UI yang berbeda peruntukkannya. Saat ini, masing-masing halaman memuat satu elemen UI)
    - App.js sekarang memuat Router, dan di dalam Router terdapat Header.js (di dalam ./src/ui), dan PrimaryLayout.js
- Websocket yang dikembangkan dari versi sebelumnya (0.4-0.5) sekarang sudah difungsikan lagi, dengan sedikit perubahan jalur, dan bentuk buffer baru.

## [1.0] - 2020-10-09

### Added
- Bentuk program yang siap menerima User Requirement baru

### Changed
- Struktur aplikasi WebServer berubah total, dikarenakan User Requirement yang berubah. Sekarang, aplikasi single page tidak memerlukan login user, dan fungsi ini dipindahkan ke halaman khusus konfigurasi.
- Struktur file berubah, di mana rencananya konten yang dimuat di router adalah yang ada di dalam ./src/pages
- Fungsi Fetcher untuk WebSocket dikembangkan dari versi sebelumnya, tetapi belum difungsikan.

## [0.5 - Unreleased]

### Changed
- Struktur file diubah sedikit. Sekarang MainWindow adalah komponen induk, berada di ./src/components, dan dimuat oleh halaman ExistBatmon.
- Perubahan struktur file membuat struktur Router (penggantian halaman berdasarkan link) juga berubah. Perubahan dapat disorot pada file :
  - ./src/ui/Header.js (komponen menu bar di bawah judul)
  - ./src/layouts/PrimaryLayout.js (layout utama yang dimuat jika user berhasil login)
  - ./src/Pages/ExistBatmon.js (sekarang komponen MainWindow ada di dalam halaman ini)

### Reminder
- Cek urutan variabel di halaman ReadCSV. Saat ini, timestamp ditempatkan pada kolom terakhir di dalam variabel dataset lokal milik halaman tersebut.
- Pikirkan cara untuk melakukan penskalaan pada variabel selain dummy. Fungsi ScaleDummy() mudah diimplementasikan karena datanya ada dalam satu chart, dan hanya chart itu saja. Dataset lain terpencar dalam tiga device yang berbeda, sehingga fungsi scaling nya perlu dibuat adaptif terhadap pilihan chart aktif.

## [0.4] - 2020-09-22

### Added
- Fitur Scaling pada bagian Eng. Unit Setup sekarang sudah difungsikan, tapi saat ini hanya untuk data dummy saja
- Kemampuan untuk menerima data lebih dari satu baterai. Versi ini dicobakan untuk data dari 2 device ADC yang terhubung dengan AI Signal Injector.

backend/node_server_v0_5.js
/************************** Import library/fungsi ****************************/
//library serialport
var serialport = require('serialport');

//library untuk server tempat buffer serta koneksi mysql
const express = require('express');

//library yang memperbolehkan CORS
const cors = require('cors');

//library untuk (format) timestamp
var moment  = require('moment');

/************************ Deklarasi objek/variabel ***************************/
//Buffer data battery monitoring dari serial
var Buff_Unit_A1 = [
    [0],[0],[0],[0],[0],
    [0],[0],
    [0],[0]
];
var timestampBuff_Unit_A1 = [
    [0],[0],[0],[0],[0],
    [0],[0],
    [0],[0]
];
var currentBuff_Unit_A1 = [
    [0],[0],[0],[0],[0],
    [0],[0],
    [0],[0]
];
const dataLimit = 20;
const normalDataLimit = 100;
// Jumlah data, atau jumlah kolom
// ada 4 tegangan 1 arus (baterai) + 2 (motor) + 2 (controller)
// 5 data + 2 data + 2 data = 9
const seriesCount = 9;
var iter = 0;

//deklarasi objek serial
var portName = "COM6";
var myPort = new serialport(portName,{
    baudRate:9600
});

//deklarasi objek server (data ke webserver)
const app = express();
const port = 5000;

//daftar request
let requests = [];

/************************ Deklarasi fungsi/event ***************************/
/*
 * 1. Data dari device ke webserver
 */
function StoreToBuffer(currentBuff,Buff,Timestamp){
    for(var i=0;i<seriesCount;i++){
        if((Buff[i].length)>dataLimit){
            Buff[i].shift();
            Buff[i][dataLimit] = currentBuff[i][0];
            Timestamp[i].shift();
            Timestamp[i][dataLimit] = String(moment().format('hh:mm:ss'));
        }
        else{
            Buff[i][Buff[i].length] = currentBuff[i][0];
            Timestamp[i][Timestamp[i].length] = String(moment().format('hh:mm:ss'));
        }
    }
    return [Buff,Timestamp];
}

//event listener untuk mengoper data dari serial ke server
function expressGETBuffer(){
    app.get('/api/buffer',(req,res) =>{
        var devices = [
            {
                id:1, name: 'Battery', 
                voltage1: Buff_Unit_A1[0],
                voltage2: Buff_Unit_A1[1],
                voltage3: Buff_Unit_A1[2],
                voltage4: Buff_Unit_A1[3],
                arus: Buff_Unit_A1[4],
                timestamp: timestampBuff_Unit_A1[0]
            },
            {
                id:2, name: 'Motor_Stepper',
                voltage1: Buff_Unit_A1[5],
                voltage2: Buff_Unit_A1[6],
                voltage3: [],
                voltage4: [],
                arus: [],
                timestamp: timestampBuff_Unit_A1[0]
            },
            {
                id:3, name: 'Controller', 
                voltage1: Buff_Unit_A1[7],
                voltage2: Buff_Unit_A1[8],
                voltage3: [],
                voltage4: [],
                arus: [],
                timestamp: timestampBuff_Unit_A1[0]
            },
        ];
        res.json(devices);
    });
    //console.log("2a. GET to Server");
}

//event Listener bila ada serial masuk dari Arduini
myPort.on("data", (line) => {
    //console.log("1a. Serial data Acquired");
    iter +=1;
    //ubah tipe data Buffer Javascript jadi Array
    bufferArray = [...line];

    //satukan data buffer menjadi integer
    V1_int = (bufferArray[1]<<8) | bufferArray[0];
    V2_int = (bufferArray[3]<<8) | bufferArray[2];
    V3_int = (bufferArray[5]<<8) | bufferArray[4];
    V4_int = (bufferArray[7]<<8) | bufferArray[6];
    Arus_int = (bufferArray[9]<<8) | bufferArray[8];
    VM1_int = (bufferArray[11]<<8) | bufferArray[10];
    VM2_int = (bufferArray[13]<<8) | bufferArray[12];
    VC1_int = (bufferArray[15]<<8) | bufferArray[14];
    VC2_int = (bufferArray[17]<<8) | bufferArray[16];

    //konversi integer ke float    
    V1 = V1_int/100.0; if(V1>normalDataLimit) V1=0;
    V2 = V2_int/100.0; if(V2>normalDataLimit) V2=0;
    V3 = V3_int/100.0; if(V3>normalDataLimit) V3=0;
    V4 = V4_int/100.0; if(V4>normalDataLimit) V4=0;
    
    Arus = Arus_int/100.0;

    VM1 = VM1_int/100.0; if(VM1>normalDataLimit) VM1=0;
    VM2 = VM2_int/100.0; if(VM2>normalDataLimit) VM2=0;
    VC1 = VC1_int/100.0; if(VC1>normalDataLimit) VC1=0;
    VC2 = VC2_int/100.0; if(VC2>normalDataLimit) VC2=0;
    //console.log("1b. Serial data Converted");

    //masukkan ke buffer agar dapat diakses chart dan server
    currentBuff_Unit_A1 = [[V1],[V2],[V3],[V4],[Arus],[VM1],[VM2],[VC1],[VC2]];
    StoreToBuffer(currentBuff_Unit_A1,Buff_Unit_A1,timestampBuff_Unit_A1);
    //console.log("1c. Serial data Stored to Buffer");
    
    console.log(bufferArray);
    //debug
    /*
    console.log(bufferArray);
    console.log(`iter:${iter}`);
    console.log(`Arus : ${Arus}`);
    console.log('V1 : '+ String(V1));
    console.log(`V2 : ${V2}`);
    */
});

myPort.on("close",() => {
    console.log("RECON coz close");
    myPort.resume();
});

myPort.on("error",() => {
    console.log("RECON coz error");
    myPort.resume();
});

/*
 * 2. Data dari webserver ke device
 */
function expressPOSTBuffer(){
    app.post('/api/request', (req, res) => {
        const newRequest = {
          userId: req.body.userId,
          title: req.body.title,
        };
      
        //requests.push(newRequest);
        console.log(newRequest);
      });
      
}
/****************************** Main Loop ***********************************/
app.use(express.json());
app.use(
    cors({
      origin: 'http://localhost:3000',
      credentials: true,
    })
  );

expressPOSTBuffer();
//untuk polling data tiap x detik
function pollData(){
    //console.log(`\n-------------------- NEW LOOP #${iter} --------------------\n`);
    pollCharacter = [0x11];
    //console.log("1. Polled Serial Port");
    myPort.write(pollCharacter);
    //console.log("2. GET Procedure");
    expressGETBuffer();
    //expressPOSTBuffer();
}
setInterval(pollData,1000);

/******************** Finalisasi (nyalakan server) *************************/
app.listen(port, () => console.log(`Server 1 started on port ${port}`));